src/librbd/api/Mirror.cc

```diff
@@ -2098,6 +2345,1689 @@ void Mirror<I>::image_snapshot_create(I *ictx, uint32_t flags,
+    // undo
+    ldout(cct, 20) << "undoing group enable: " << ret_code << dendl;
+    remove_interim_snapshots(group_ioctx, group_header_oid,
+                             &image_ctxs, &group_snap);
+    for (size_t i = 0; i < image_ctxs.size(); i++) {
+      if (snap_ids[i] == CEPH_NOSNAP) {
+        continue;
+      }
+      r = image_disable(image_ctxs[i], false);
+      if (r < 0) {
+        lderr(cct) << "failed to disable mirroring on image: "
+                   << image_ctxs[i]->name << cpp_strerror(r) << dendl;
+      }
+    }
+  } else {
+    mirror_group.state = cls::rbd::MIRROR_GROUP_STATE_ENABLED;
+    r = cls_client::mirror_group_set(&group_ioctx, group_id, mirror_group);
+    if (r < 0) {
+      lderr(cct) << "failed to update mirroring group metadata: "
+                 << cpp_strerror(r) << dendl;
+      ret_code = r;
+    }
+
+    group_snap.state = cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE;
+    r = cls_client::group_snap_set(&group_ioctx, group_header_oid, group_snap);
+    if (r < 0) {
+      lderr(cct) << "failed to update group snapshot metadata: "
+                 << cpp_strerror(r) << dendl;
+    }
+  }
+
+cleanup:
+  if (!quiesce_requests.empty()) {
+    util::notify_unquiesce(image_ctxs, quiesce_requests);
+  }
+
+  close_images(&image_ctxs);
+
+  if (!ret_code) {
+    r = MirroringWatcher<I>::notify_group_updated(
+        group_ioctx, cls::rbd::MIRROR_GROUP_STATE_ENABLED, group_id,
+        mirror_group.global_group_id, image_count);
+    if (r < 0) {
+      lderr(cct) << "failed to notify mirroring group updated: "
+                 << cpp_strerror(r) << dendl;
+      // not fatal
+    }
+  }
+
+  return ret_code;
+}
+
+template <typename I>
+int Mirror<I>::group_disable(IoCtx& group_ioctx, const char *group_name,
+                            bool force) {
+  CephContext *cct = (CephContext *)group_ioctx.cct();
+  ldout(cct, 20) << "io_ctx=" << &group_ioctx
+		 << " group name " << group_name
+		 << " force " << force << dendl;
+
+  std::string group_id;
+  int r = cls_client::dir_get_id(&group_ioctx, RBD_GROUP_DIRECTORY,
+				 group_name, &group_id);
+  if (r < 0) {
+    lderr(cct) << "error reading group id object: "
+               << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  cls::rbd::MirrorGroup mirror_group;
+  r = cls_client::mirror_group_get(&group_ioctx, group_id, &mirror_group);
+  if (r == -EOPNOTSUPP) {
+    ldout(cct, 5) << "group mirroring not supported by OSD" << dendl;
+    return r;
+  } else if (r == -ENOENT) {
+    ldout(cct, 10) << "mirroring for group " << group_name
+                   << " already disabled" << dendl;
+    return 0;
+  } else if (r < 0) {
+    lderr(cct) << "failed to retrieve mirror group metadata: "
+               << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  std::vector<I *> image_ctxs;
+  r = open_group_images(group_ioctx, group_id, &image_ctxs);
+  if (r < 0) {
+    return r;
+  }
+
+  mirror_group.state = cls::rbd::MIRROR_GROUP_STATE_DISABLING;
```

Comment:
Check if the group is primary and force is false before proceeding to diable images.